### PR TITLE
Add support for generating TypeScript declarations, and named pipelines

### PR DESCRIPTION
**Pull request by devongovett**
This adds support for generating TypeScript declaration files as a library target, along with the beginnings of named pipeline support. This allows the same input file to be sent through multiple different pipelines to produce different outputs (a `.js` file and a `.d.ts` file in this case). A `pipeline` parameter has been added to dependencies, asset requests, and config requests to support this in core. Related: #3477.

Eventually, resolvers should be able to return pipeline names parsed from import specifiers, so you could reference pipelines from code. For now, the pipeline is set on entries based on the target name (e.g. `types` from package.json). So, you can have a package.json like this:

```json
{
  "name": "my-package",
  "main": "dist/index.js",
  "types": "dist/index.d.ts"
}
```

And a `.parcelrc` like this (actually, this is in the default config):

```json
{
  "transforms": {
    "types:*.{ts,tsx}": "@parcel/transformer-typescript-types",
  }
}
```

Parcel will resolve the targets in package.json, and apply the appropriate pipeline based on the target name. If no named pipeline is found for that target (e.g. no `main` pipeline), then it will fall back to the non-named pipeline that matches.

In terms of actually generating TS declarations, we use the TypeScript compiler and a custom TypeScript transformer to do that. TypeScript normally generates separate `.d.ts` file for each file that you import in your code, but if you specify the `outFile` option it will generate a "bundle" file instead with all modules inlined. Unfortunately, this still includes all types, not only ones that are exported by the package. 

So, we have a custom TS transformer which uses the AST to tree shake and flatten these modules into one top-level `.d.ts` file containing only exported types. It works in three passes, similar to the JS tree shaking implementation: first collecting information about the imported, exported, and local declarations in each module, then propagating exports from the entry module through to the children and marking used nodes, and finally removing unused nodes and flattening into the top-level scope, renaming declarations as needed.

### CODE REVIEW

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>packages/benchmarks/kitchen-sink</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 11.20s | -473.00ms |
| Cached | 8.98s | -89.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/modern/index.js | 794.00b | +0.00b | 541.00ms | -28.00ms 🚀 |
| dist/modern/index.css | 29.00b | +0.00b | 1.40s | +74.00ms ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/index.js | 252.00b | +0.00b | 16.00ms | -2.00ms 🚀 |
| dist/modern/index.css | 29.00b | +0.00b | 11.00ms | -1.00ms 🚀 |

</p></details>

<details><summary>packages/benchmarks/react-hn</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 15.99s | +188.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/245584963556082186)**

File: packages/shared/scope-hoisting/src/link.js [hunk 1]
```diff
@@ -545,11 +545,10 @@ export function link({
...
 
-        let paths = path.unshiftContainer('body', imports);
-        for (let path of paths) {
-          if (path.isDeclaration()) {
-            path.scope.registerDeclaration(path);
-          }
+        if (imports.length > 0) {
+          // Add import statements and update scope to collect references
+          path.unshiftContainer('body', imports);
+          path.scope.crawl();
```
Comment (mischnic) [inline packages/shared/scope-hoisting/src/link.js]: Can you a test case for that? (So it doesn't break again if we come back to optimize this)

Comment (devongovett) [reply packages/shared/scope-hoisting/src/link.js]: done!

File: packages/core/core/src/Dependency.js [hunk 2]
```diff
@@ -31,7 +32,8 @@ export function createDependency(opts: DependencyOpts): Dependency {
     md5FromString(
```
Comment (wbinnssmith) [inline packages/core/core/src/Dependency.js]: there are a lot of things in here now. what about using `md5FromObject`? something like

```
md5FromObject({
  sourceAssetId: opts.sourceAssetId,
  moduleSpecifier: opts.moduleSpecifier,
  env: opts.env,
  target: opts.target,
  pipeline: opts.pipeline
})

File: packages/packagers/ts/src/TSPackager.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
+// @flow strict-local
```
Comment (wbinnssmith) [inline packages/packagers/ts/src/TSPackager.js]: Is this only for packaging type definitions? Maybe the name of the file could include "types"?

File: packages/configs/default/index.json [hunk 4]
```diff
@@ -1,6 +1,7 @@
 {
   "bundler": "@parcel/bundler-default",
   "transforms": {
+    "types:*.{ts,tsx}": ["@parcel/transformer-typescript-types"],
```
Comment (jamiebuilds) [inline packages/configs/default/index.json]: Is this supposed to be the syntax for named pipelines?

Comment (devongovett) [reply packages/configs/default/index.json]: Yes.

**Pull request merged**